sql/wsrep_mysqld.h

```diff
@@ -369,7 +369,104 @@ class Log_event;
 int wsrep_ignored_error_code(Log_event* ev, int error);
 int wsrep_must_ignore_error(THD* thd);
 
-bool wsrep_replicate_GTID(THD* thd);
+extern bool wsrep_gtid_mode;
+typedef struct
+{
+  uint32 domain_id;
+  uint32 server_id;
+  uint64 seqno;
+} wsrep_server_gtid_t;
+typedef struct
+{
+  bool found;
+  bool valid;
+  wsrep_server_gtid_t gtid;
+} wsrep_server_gtid_wait_t;
+class Wsrep_gtid_server
+{
+public:
+  uint32 domain_id;
+  uint32 server_id;
+  void gtid(const wsrep_server_gtid_t& gtid)
+  {
+    domain_id=  gtid.domain_id;
+    server_id=  gtid.server_id;
+    m_seqno=    gtid.seqno;
+  }
+  wsrep_server_gtid_t gtid()
+  {
+    wsrep_server_gtid_t gtid;
+    gtid.domain_id= domain_id;
+    gtid.server_id= server_id;
+    gtid.seqno=     m_seqno;
+    return gtid;
+  }
+  void seqno(const uint64 seqno) { m_seqno= seqno; }
+  uint64 seqno() const { return m_seqno; }
+  uint64 seqno_inc() 
+  {
+    m_seqno++;
+    return m_seqno;
+  }
+  const wsrep_server_gtid_t& undefined()
+  {
+    return m_undefined;
+  }
+  int wait_gtid_upto(wsrep_server_gtid_wait_t* gtid_wait_list, uint count,
+                     uint timeout)
+  {
+    int wait_result;
+    uint gtid_count= count, i;
+    struct timespec wait_time;
+    int ret= 0;
+    mysql_mutex_lock(&LOCK_wsrep_gtid_wait_upto);
+    set_timespec(wait_time, timeout);
+    for (i= 0; i < count && !gtid_wait_list[i].valid; ++i) gtid_count--;
+    while (1)
```

Comment:
This while loop may lead to "Thundering Herd Problem" (https://en.wikipedia.org/wiki/Thundering_herd_problem) if there  are many waiters. What is worse, the condition wait inside the loop returns every time the `COND_wsrep_gtid_wait_upto` is signaled, which may happen several times for single wait. Also the wait happens before GTIDs are inspected, so this wait will fail every time the signal from `set_last_written_gtid()` is delivered before the waiter reaches `mysql_cond_timedwait()`.

Reply:
Sync wait upto is now refactored to address problems mentioned in this review.